src/include/mempool.h

```diff
@@ -0,0 +1,521 @@
+
+   unittest_1::map
+   unittest_1::multimap
+   unittest_1::set
+   unittest_1::multiset
+   unittest_1::list
+   unittest_1::vector
+   unittest_1::unordered_map
+
+
+Putting objects in a mempool
+----------------------------
+
+In order to use a memory pool with a particular type, a few additional
+declarations are needed.
+
+For a class:
+
+  struct Foo {
+    MEMBER_OF_MEMPOOL()
+    ...
+  };
+
+Then, in an appropriate .cc file,
+
+  MEMPOOL_DEFINE_OBJECT_FACTORY(Foo, foo, unittest_1);
+
+The second argument can generally be identical to the first, except
+when the type contains a nested scope.  For example, for
+BlueStore::Onode, we need to do
+
+  MEMPOOL_DEFINE_OBJECT_FACTORY(BlueStore::Onode, bluestore_onode,
+                                bluestore_meta);
+
+(This is just because we need to name some static varables and we
+can't use :: in a variable name.)
+
+In order to use the STL containers, a few additional declarations
+are needed.  For example,
+
+  unittest_1::map<int> myvec;
+
+requires
+
+  MEMPOOL_DEFINE_FACTORY(int, int, unittest_1);
+  MEMPOOL_DEFINE_MAP_FACTORY(int, int, unittest_1);
+
+There are similar macros for SET, LIST, and UNORDERED_MAP.  The MAP
+macro serves both std::map and std::multimap, and std::vector doesn't
+need a second container-specific declaration (because it only
+allocates T and T* arrays; there is no internal container-specific
+wrapper type).
+
+unordered_map is trickier.  First, it has to allocate the hash
+buckets, which requires an extra mempool-wide definition that is
+shared by all different types.  Second, sometimes the hash value is
+cached in the hash node and sometimes it is not.  The glibc STL makes
+its own decision if you don't explicitly define traits, so you either
+need to match your definition with its inference, or explicitly define
+traits, or simply define the allocator for with the cached and
+uncached case and leave one of them unused (but polluting your
+debug dumps).  For example,
+
+  unittest_2::unordered_map<int, std::string> one;       // not cached
+  unittest_2::unordered_map<std::string, int> one;       // cached
+
+needs
+
+ MEMPOOL_DEFINE_UNORDERED_MAP_BASE_FACTORY(unittest_2);  // used by both
+ MEMPOOL_DEFINE_UNORDERED_MAP_FACTORY(uint64_t, std::string, false, int_str,
+                                      unittest_2);
+ MEMPOOL_DEFINE_UNORDERED_MAP_FACTORY(std::string, uint64_t, true, str_int,
+                                      unittest_2);
+
+
+Introspection
+-------------
+
+The simplest way to interrogate the process is with
+
+  Formater *f = ...
+  mempool::dump(f);
+
+This will dump information about *all* memory pools.
+
+You can also interogate a specific pool programatically with
+
+  size_t bytes = unittest_2::allocated_bytes();
+  size_t items = unittest_2::allocated_items();
+
+Note that you cannot easily query per-type, primarily because debug
+mode is optional and you should not rely on that information being
+available.
+
```

Comment:
Should include a comment about the run-time complexity of these interrogation calls.